lib/vega_lite/data.ex

```diff
@@ -0,0 +1,231 @@
+
+  @doc """
+  Same as chart/3 but receives a valid `VegaLite` specification as a first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "With title")
+      |> Data.chart(data, :bar, x: "category", y: "score")
+
+      Vl.new(title: "With title")
+      |> Vl.mark(:bar)
+      |> Data.chart(data, x: "category", y: "score")
+
+  The above example achieves the same results as the example below.
+
+      Vl.new(title: "With title")
+      |> Vl.data_from_values(data, only: ["category", "score"])
+      |> Vl.mark(:bar)
+      |> Vl.encode_field(:x, "category", type: :nominal)
+      |> Vl.encode_field(:y, "score", type: :quantitative)
+  """
+  @spec chart(VegaLite.t(), Table.Reader.t(), atom() | keyword(), keyword()) :: VegaLite.t()
+  def chart(vl, data, mark, fields) do
+    cols = columns_for(data)
+    used_fields = fields |> Keyword.values() |> used_fields()
+    root = vl |> Vl.data_from_values(data, only: used_fields) |> encode_mark(mark)
+    build_fields(fields, root, cols)
+  end
+
+  @doc """
+  Returns the specification of a heat map for a given data and a list of fields to be encoded.
+
+  As a specialized chart, the heatmap expects an `:x` and `:y` and optionally a `:color` and a
+  `:text` field. Defaults to `:nominal` for the axes and `:quantitative` for color and text if
+  types are not specified.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+
+  """
+  @spec heatmap(Table.Reader.t(), keyword()) :: VegaLite.t()
+  def heatmap(data, fields), do: heatmap(Vl.new(), data, fields)
+
+  @doc """
+  Same as heatmap/2, but takes a valid `VegaLite` specification as the first argument.
+
+  ## Examples
+
+      data = [
+        %{"category" => "A", "score" => 28},
+        %{"category" => "B", "score" => 55}
+      ]
+
+      Vl.new(title: "Heatmap", width: 500)
+      |> Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
+  """
+  @spec heatmap(VegaLite.t(), Table.Reader.t(), keyword()) :: VegaLite.t()
+  def heatmap(vl, data, fields) do
+    fields = standardize_fields(fields) |> Enum.map(&heatmap_defaults/1)
+
+    case {fields[:x], fields[:y], fields[:text]} do
+      {nil, _, _} -> raise ArgumentError, "the x axis is required to plot a heatmap"
+      {_, nil, _} -> raise ArgumentError, "the y axis is required to plot a heatmap"
+      {_, _, nil} -> chart(vl, data, :rect, fields)
+      _ -> annotated_heatmap(vl, data, fields)
+    end
+  end
+
+  defp annotated_heatmap(vl, data, fields) do
+    text_fields = [text: fields[:text], x: fields[:x], y: fields[:y]]
+    fields = Keyword.delete(fields, :text)
+    used_fields = fields |> Keyword.values() |> used_fields()
```

Comment:
I think there is a bug here, no? It seems we will never include `:text` in `used_fields`. 

```suggestion
    rect_fields = Keyword.delete(fields, :text)
    used_fields = fields |> Keyword.values() |> used_fields()
```

Also, my suggestion is to remove `Keyword.values()`. Pass the whole fields to `used_fields()` and you can just ignore the `key` there.